### PR TITLE
chore: rework versioning

### PR DESCRIPTION
**Pull request by henryiii**
Rework versioning to come from pyproject.toml. ~~Other cleanups.~~

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -69,7 +69,7 @@ The following platforms are supported with binary wheels:
   | macOS 11+     | Apple Silicon            |
   +---------------+--------------------------+
 
-The last version to provide ``manylinux1`` wheels was ``3.22.x``.
+The last version to provide ``manylinux1`` wheels was ``3.22.x``. The last version to provide Python 2 to Python 3.6 support was ``3.28.x``.
```
Comment (jcfr) [inline README.rst:72]:
```suggestion
The last version to provide ``manylinux1`` wheels was ``3.22.x``.
The last version to provide Python 2 to Python 3.6 support was ``3.28.x``.
```

File: pyproject.toml [hunk 2]
```diff
@@ -50,73 +50,72 @@ ctest = "cmake:ctest"
...
-]
-environment = { CMAKE_ARGS = "-DOPENSSL_ROOT_DIR:PATH=/usr/local/ssl -DCMAKE_JOB_POOL_COMPILE:STRING=compile -DCMAKE_JOB_POOL_LINK:STRING=link -DCMAKE_JOB_POOLS:STRING=compile=4;link=1 -DCMAKE_CXX_STANDARD:STRING=11" }
+before-all = "./scripts/manylinux-build-and-install-openssl.sh"
 
-[[tool.cibuildwheel.overrides]]
-select = ["*-manylinux_aarch64", "*-manylinux_ppc64le", "*-manylinux_s390x"]
-# disable tests on those platforms, QEMU is taking to long for jobs to pass on GHA
-environment = { CMAKE_ARGS = "-DOPENSSL_ROOT_DIR:PATH=/usr/local/ssl -DCMAKE_JOB_POOL_COMPILE:STRING=compile -DCMAKE_JOB_POOL_LINK:STRING=link -DCMAKE_JOB_POOLS:STRING=compile=4;link=1 -DCMAKE_CXX_STANDARD:STRING=11 -DRUN_CMAKE_TEST:BOOL=OFF" }
+[tool.cibuildwheel.linux.config-settings]
+"cmake.define.OPENSSL_ROOT_DIR" = "/usr/local/ssl"
```
Comment (jcfr) [inline pyproject.toml:78]: I suggest to contribute these updates in a dedicated PR

File: CMakeLists.txt [hunk 3]
```diff
@@ -33,7 +33,13 @@ if(CMakePythonDistributions_SUPERBUILD)
...
   option(BUILD_VERBOSE "Build reporting additional information (e.g download progress, ...)" ON)
 
-  option(RUN_CMAKE_TEST "Run CMake test suite when built from sources" ON)
+  if("$ENV{CIBUILDWHEEL}")
+    set(CIBW ON)
+  else()
+    set(CIBW OFF)
+  endif()
+
+  option(RUN_CMAKE_TEST "Run CMake test suite when built from sources" "${CIBW}")
```
Comment (jcfr) [inline CMakeLists.txt:42]: Is there a way to set the `RUN_CMAKE_TEST` option in the corresponding `tool.cibuildwheel` section in `pyproject.toml` ?

File: pyproject.toml [hunk 4]
```diff
@@ -4,6 +4,7 @@ build-backend = "scikit_build_core.build"
 
 [project]
 name = "cmake"
+version = "3.29.0"
```
Comment (jcfr) [inline pyproject.toml:7]: When we need to created patch release like `3.29.0.1`[^1], I am assuming we will update this entry accordingly ?

[^1]: https://github.com/scikit-build/cmake-python-distributions/tree/3.29.0.1

Comment (henryiii) [reply pyproject.toml:7]: Yes

File: scripts/update_cmake_version.py [hunk 5]
```diff
@@ -172,24 +169,29 @@ def update_docs(version):
...
     _update_file(
         os.path.join(ROOT_DIR, "tests/test_cmake.py"), pattern, replacement
     )
 
 
+def update_pyproject_toml(version):
+    pattern = re.compile(r'version = "[\w\.]+"')
+    replacement = 'version = "%s"' % version
+    _update_file(
+        os.path.join(ROOT_DIR, "tests/test_cmake.py"), pattern, replacement
```
Comment (jcfr) [inline scripts/update_cmake_version.py:189]:
```suggestion
        os.path.join(ROOT_DIR, "pyproject.toml"), pattern, replacement
```

File: scripts/update_cmake_version.py [hunk 6]
```diff
@@ -172,24 +169,29 @@ def update_docs(version):
...
-    pattern = re.compile(r'expected_version = "\d.(\d)+.\d"')
+    pattern = re.compile(r'expected_version = "\d.\d+.\d"')
     replacement = 'expected_version = "%s"' % version
     _update_file(
         os.path.join(ROOT_DIR, "tests/test_cmake.py"), pattern, replacement
     )
 
 
+def update_pyproject_toml(version):
+    pattern = re.compile(r'version = "[\w\.]+"')
```
Comment (jcfr) [inline scripts/update_cmake_version.py:186]:
```suggestion
    pattern = re.compile(r'^version = "[\w\.]+"$')
```

File: noxfile.py [hunk 7]
```diff
@@ -157,3 +158,16 @@ def bump_openssl(session: nox.Session) -> None:
...
+
+@nox.session(venv_backend="none")
+def tag_release(session: nox.Session) -> None:
+    """
+    Print instructions for tagging a release and pushing it to GitHub.
+    """
+
+    session.log("Run the following commands to make a release:")
+    txt = Path("pyproject.toml").read_text()
+    current_version = next(iter(re.finditer(r'version = "([\d\.]+)"', txt))).group(1)
```
Comment (jcfr) [inline noxfile.py:171]:
```suggestion
    current_version = next(iter(re.finditer(r'^version = "([\d\.]+)$"', txt))).group(1)
```

File: docs/conf.py [hunk 8]
```diff
@@ -61,11 +65,11 @@
 # The version info for the project you're documenting, acts as replacement for
 # |version| and |release|, also used in various other places throughout the
 # built documents.
-#
-# The short X.Y version.
-version = metadata.version("cmake")
-# The full version, including alpha/beta/rc tags.
-release = version
+with pyproject.open('rb') as f:
+    release = tomllib.load(f)["project"]["version"]
```
Comment (henryiii) [inline docs/conf.py:69]: This fixes the docs build, which has been broken for quite a while.

**Comment by thewtex** [comment]
Discussion note:

Add *pyproject.toml* to the nox `bump` function.

**Review by thewtex — COMMENTED** [review]
LGTM

**Comment by henryiii** [comment]
> Add pyproject.toml to the nox bump function.

Thanks! I'd already forgotten.

**Pull request merged**